metals/src/main/scala/scala/meta/internal/metals/FileDecoderProvider.scala

```diff
@@ -0,0 +1,285 @@
+import scala.meta.internal.metap.Main
+import scala.meta.internal.mtags.SemanticdbClasspath
+import scala.meta.io.AbsolutePath
+import scala.meta.metap.Format
+import scala.meta.metap.Settings
+import scala.util.Properties
+
+final class FileDecoderProvider(
+    workspace: AbsolutePath,
+    buildTargets: BuildTargets,
+    userConfig: () => UserConfiguration,
+    shellRunner: ShellRunner
+)(implicit ec: ExecutionContext) {
+
+  private sealed abstract class Error(val message: String)
+  private sealed abstract class ErrorWithException(
+      message: String,
+      val ex: Throwable
+  ) extends Error(message)
+  private case class InvalidURI(uri: String, override val ex: Throwable)
+      extends ErrorWithException(s"Invalid URI `$uri`", ex)
+  private case class InvalidScheme(uri: URI)
+      extends Error(s"Invalid scheme `${uri.getScheme}`")
+  private case class InvalidQuery(uri: URI)
+      extends Error(s"Invalid query `${uri.getQuery}`")
+  private case class InvalidQueryParams(
+      params: Array[Array[String]],
+      override val ex: Throwable
+  ) extends ErrorWithException(
+        s"Invalid parameters `${params.map(_.mkString("=")).mkString(",")}`",
+        ex
+      )
+  private case class MissingDecoder(uri: URI)
+      extends Error(s"Missing `decoder` argument `$uri`")
+  private case class MissingOriginalURI(uri: URI)
+      extends Error(s"Missing `originalURI` argument `$uri`")
+  private case class InvalidURIPath(path: String, override val ex: Throwable)
+      extends ErrorWithException(s"Invalid URI path `$path`", ex)
+  private case class FileMissing(path: AbsolutePath)
+      extends Error(s"File missing `$path`")
+  private case class InvalidDecoder(decoder: String)
+      extends Error(s"Invalid `$decoder`")
+  private case class FileToDecodeMissing(path: AbsolutePath)
+      extends Error(s"""|No file to decode for $path
+                        |Is Metals still starting?
+                        |Is the file compiled?
+                        |""".stripMargin)
+  private case class DecoderError(
+      path: AbsolutePath,
+      override val ex: Throwable
+  ) extends ErrorWithException(s"Error decoding file `$path`", ex)
+
+  private case class Decoder(
+      findFile: AbsolutePath => Option[AbsolutePath],
+      decode: AbsolutePath => Future[Either[Error, String]]
+  )
+
+  /**
+   * URI format...
+   * metalsDecodeFile:/fileUrl?decoder=decoderName&keyValueArguments&originalURI=originalFileURI
+   *
+   * Examples...
+   * metalsDecodeFile:/somedir/someFile.java.javap.verbose?decoder=javap&originalURI=file:///somedir/someFile.java
+   * metalsDecodeFile:/somedir/someFile.class.javap.verbose?decoder=javap&originalURI=file:///somedir/someFile.class
+   * metalsDecodeFile:/somedir/someFile.scala?decoder=semanticdb&originalURI=file:///somedir/someFile.scala
+   * metalsDecodeFile:/somedir/someFile.scala.semanticdb?decoder=semanticdb&originalURI=file:///somedir/someFile.scala.semanticdb
```

Comment:
Do we need originalURI?
We could just get the relevant parts from the existing uri. Set query to none and scheme to file.

Reply:
I was wondering about the potential for a file in a jar.  isn't the uri for that something like "jar:file:/path/file.class"?  And I saw some chatter on Discord about using read-only files which VirtualDocuments provides so maybe this could be extended to navigate to jar sources and decompile classes in jars that don't have sources?

Reply:
We never show class files, so currently we would never send a jar URI. But potentially we could try returning it in the future when sources are not available and decompile it, but in that case we would don't really need the source file, since we can assume class files can be decompiled with javap